genesyscloud/user/genesyscloud_user_proxy.go

```diff
@@ -0,0 +1,247 @@
+
+	//Inner function to get user based on status
+	getUsersByStatus := func(userStatus string) (*[]platformclientv2.User, *platformclientv2.APIResponse, error) {
+		users := []platformclientv2.User{}
+		const pageSize = 100
+
+		usersList, apiResponse, err := p.userApi.GetUsers(pageSize, 1, nil, nil, "", nil, "", userStatus)
+		if err != nil {
+			return nil, apiResponse, err
+		}
+		users = append(users, *usersList.Entities...)
+
+		for pageNum := 2; pageNum <= *usersList.PageCount; pageNum++ {
+			usersList, _, err := p.userApi.GetUsers(pageSize, pageNum, nil, nil, "", nil, "", userStatus)
+			if err != nil {
+				return nil, apiResponse, err
+			}
+			users = append(users, *usersList.Entities...)
+		}
+
+		return &users, apiResponse, nil
+	}
+
+	// Get all "active" and "inactive" users
+	allUsers := []platformclientv2.User{}
+
+	activeUsers, apiResponse, err := getUsersByStatus("active")
+	if err != nil {
+		return nil, apiResponse, fmt.Errorf("failed to get 'active' users %v", err)
+	}
+	allUsers = append(allUsers, *activeUsers...)
+
+	inactiveUsers, apiResponse, err := getUsersByStatus("inactive")
+	if err != nil {
+		return nil, apiResponse, fmt.Errorf("failed to get 'inactive' users %v", err)
+	}
+	allUsers = append(allUsers, *inactiveUsers...)
+
+	// Cache the architect schedules resource into the p.userCache for later use
+	for _, user := range allUsers {
+		rc.SetCache(p.userCache, *user.Id, user)
+	}
+
+	return &allUsers, apiResponse, nil
+}
+
+// getUserIdByNameFn is an implementation of the function to get a Genesys Cloud user by name
+func getUserIdByNameFn(ctx context.Context, p *userProxy, name string) (id string, retryable bool, response *platformclientv2.APIResponse, err error) {
+	users, apiResponse, err := getAllUserFn(ctx, p)
+	if err != nil {
+		return "", false, apiResponse, err
+	}
+
+	if users == nil || len(*users) == 0 {
+		return "", false, apiResponse, fmt.Errorf("No User found with name %s", name)
+	}
+
+	for _, user := range *users {
+		if *user.Name == name {
+			log.Printf("Retrieved the user id %s by name %s", *user.Id, name)
+			return *user.Id, false, apiResponse, nil
+		}
+	}
+
+	return "", true, apiResponse, fmt.Errorf("Unable to find user wiht name %s", name)
+}
+
+// getUserByIdFn is an implementation of the function to get a Genesys Cloud user by Id
+func getUserByIdFn(ctx context.Context, p *userProxy, id string, expand []string, state string) (user *platformclientv2.User, response *platformclientv2.APIResponse, err error) {
+	userLocal, apiResponse, err := p.userApi.GetUser(id, expand, "", state)
+	if err != nil {
+		return nil, apiResponse, fmt.Errorf("Failed to retrieve user by id %s: %s", id, err)
+	}
+	return userLocal, apiResponse, nil
+}
+
+// updateUserFn is an implementation of the function to update a Genesys Cloud user
+func updateUserFn(ctx context.Context, p *userProxy, id string, updateUser *platformclientv2.Updateuser) (*platformclientv2.User, *platformclientv2.APIResponse, error) {
```

Comment:
For some of the functions like this where we just call the api, check the error and return the error if exists or the response body if it doesn't. You can just return the api call since we check the error in the calling function anyway. So this function could just be:
```
return p.userApi.PatchUser(id, *updateUser)
```

Reply:
I think not only to this function it applies to few other functions as well in the proxy. Also I believe that we have to do this code optimization in several other packages :)

Reply:
Correct, we started with this pattern of checking the error before returning but eventually someone realised it was unnecessary